File: 0000-0999/0212.WordSearchII.go
func findWords(board [][]byte, words []string) []string {
    ans := []string{}
    root := buildTrie(words)
    for i := 0; i < len(board); i++ {
        for j := 0; j < len(board[0]); j++ {
            dfs(board, i, j, root, &ans)
        }
    }
    return ans
}

func dfs(board [][]byte, i, j int, node *TrieNode, ans *[]string) {
    if node.word != "" {
        *ans = append(*ans, node.word)
        node.word = ""
    }
    if i < 0 || i >= len(board) || j < 0 || j >= len(board[0]) || board[i][j] == '#' || node.child[board[i][j]-'a'] == nil {
        return
    }
    tmp := board[i][j]
    board[i][j] = '#'
    node = node.child[tmp-'a']
    dfs(board, i-1, j, node, ans)
    dfs(board, i+1, j, node, ans)
    dfs(board, i, j+1, node, ans)
    dfs(board, i, j-1, node, ans)
    board[i][j] = tmp
}

type TrieNode struct {
    child [26]*TrieNode
    word string
}

func buildTrie(words []string) *TrieNode {
    root := &TrieNode{}
    for _, w := range words {
        node := root
        for _, ch := range w {
            if node.child[ch-'a'] == nil {
                node.child[ch-'a'] = &TrieNode{}
            }
            node = node.child[ch-'a']
        }
        node.word = w
    }
    return root
}

File: 0000-0999/0212.WordSearchII.py
class TrieNode(object):
    def __init__(self):
        self.children = {}
        self.isWord = False

class Trie(object):
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
        node.isWord = True

    def search(self, word):
        node = self.root
        for ch in word:
            if ch not in node.children:
                return False
            node = node.children[ch]
        return node.isWord

class Solution(object):
    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        ans = []
        trie = Trie()
        for word in words:
            trie.insert(word)
        for i in xrange(len(board)):
            for j in xrange(len(board[0])):
                self.dfs(board, trie.root, i, j, '', ans)
        return ans

    def dfs(self, board, node, i, j, path, ans):
        if node.isWord:
            ans.append(path)
            node.isWord = False
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]):
            return
        temp = board[i][j]
        if temp not in node.children:
            return
        board[i][j] = '#'
        node = node.children[temp]
        self.dfs(board, node, i-1, j, path+temp, ans)
        self.dfs(board, node, i+1, j, path+temp, ans)
        self.dfs(board, node, i, j-1, path+temp, ans)
        self.dfs(board, node, i, j+1, path+temp, ans)
        board[i][j] = temp
